File: script/BugBase/bin/run.bugbase.web.r
# Runs BugBase analysis:
# - Uses a single-cell approach to predict microbiome phenotypes
# - plots phenotype relative abundances
# - prints statisical analyses
# - plots thresholds
# - plots otu contributions

# USAGE
# Default
# predict_phenotypes.r 	-i otu_table.txt 
#						-m mapping_file.txt 
#						-c map_column 
#						-o output_directory_name 

# Options
# -t 	Specify which taxa level to plot otu contributions by (number 1-7) 
# -p 	Specify list of traits (phenotypes) to test (list, comma separated)
# -T 	Specify a threshold to use for all traits (number 0-1)
# -g 	Specify subset of groups in map column to plot (list, comma separated)
# -u  Use a user-define trait table. Absolute file path must be specified
# -k  Use the kegg pathways instead of default traits
# -z 	Data is of type continuous 
# -C 	Use covariance instead of variance to determine thresholds
# -a 	Plot all samples (no stats will be run)

options(warn = -1)

# Set the enivornment paths:
my_env <- '/web/research/bugbase.cs.umn.edu/BugBase'
lib_location <- '/web/research/bugbase.cs.umn.edu/R_lib'
#uploads <- '/web/research/bugbase.cs.umn.edu/uploads'
#results <- '/web/research/bugbase.cs.umn.edu/results'

# Load packages
library(optparse, lib.loc = lib_location)
library(reshape2, lib.loc = lib_location) 
library(plyr, lib.loc = lib_location)
library(RColorBrewer, lib.loc = lib_location)
library(gridExtra, lib.loc = lib_location)
library(ggplot2, lib.loc = lib_location)
library(beeswarm, lib.loc = lib_location)
library(biom, lib.loc = lib_location)
library(RJSONIO, lib.loc= lib_location) #
#library(Matrix, lib.loc= lib_location) #
library(labeling, lib.loc = lib_location)
library(digest, lib.loc = lib_location) #
library(methods)

# Find functions in the lib
lib_dir <- paste(my_env, "/lib", sep='')
r_funcs <- list.files(lib_dir)
for (i in 1:length(r_funcs)){
	source(paste(lib_dir, r_funcs[[i]], sep = "/"))
}

option_list <- list(
  make_option(c("-v", "--verbose"), action="store_true", default=TRUE,
              help="Print extra output [default]"),
  make_option(c("-i", "--otutable"), type="character", default=NULL,
              help="otu table to plot [default %default]"),
  make_option(c("-c", "--mapcolumn"), type="character", default=NULL,
              help="column of mapping file to plot [default %default]"),
  make_option(c("-m", "--mappingfile"), type="character", default=NULL,
              help="mapping file to plot [default %default]"),
  make_option(c("-o", "--output"), type="character", default=".",
              help="output directory [default %default]"),
  make_option(c("-t", "--taxalevel"), type="character", default=NULL,
              help="taxa level to plot otu contributions by, default is 
              2 (phylum) [default %default]"),
  make_option(c("-p", "--phenotype"), type="character", default=NULL,
              help="specific traits (phenotypes) to predict, separated by 
              commas, no spaces [default %default]"),
  make_option(c("-x", "--predict"), action="store_true", default=FALSE,
              help="only output the prediction table, do not make plots 
              [default %default]"),
  make_option(c("-T", "--threshold"), type="character", default=NULL,
              help="threshold to use, must be between 0 and 1 
              [default %default]"),
  make_option(c("-g", "--groups"), type="character", default=NULL,
              help="treatment groups of samples, separated by commas, no spaces 
              [default %default]"),
  make_option(c("-u", "--usertable"), type="character", default=NULL,
              help="user define trait table, absolute file path required [
              default %default]"),
  make_option(c("-z", "--continuous"), action="store_true", default=FALSE,
              help="plot continuous data [default %default]"),
  make_option(c("-k", "--kegg"), action="store_true", default=FALSE,
              help="use kegg pathway table [default %default]"),
  make_option(c("-C", "--cov"), action="store_true", default=FALSE,
              help="use coefficient of variance instead of variance [default %default]"),
  make_option(c("-l", "--clr_trans"), action="store_true", default=FALSE,
              help="use centered log-ratio transformation instead of relative abundance [default %default]"),
  make_option(c("-a", "--all"), action="store_true", default=FALSE,
              help="plot all samples without a mapping file (this outputs no 
              statistics) [default %default]"),
  make_option(c("-w", "--shotgun"), action="store_true", default=FALSE,
              help="Data is metagenomic shotgun data 
              (picked against RefSeq database) [default %default]")
  )
opts <- parse_args(OptionParser(option_list=option_list))

# Define the database files
db_fp <- paste(my_env, "/usr", sep='')

#Check for shotgun and KEGG
if(isTRUE(opts$shotgun)){
  copy_no_file <- paste(db_fp, "16S_13_5_precalculated.txt.gz", sep='/')
  taxonomy <- paste(db_fp, "img_otu_taxonomy.txt.gz", sep='/')
  if(isTRUE(opts$kegg)){
    trait_table <- paste(db_fp, 
    	"kegg_modules_img_precalculated.txt.gz", sep='/')
    if(!isTRUE(opts$predict)){
      if(is.null(opts$phenotype)){
      stop("A list of modules must be specified when using KEGG")
      }
    }
  } else {
    if(is.null(opts$usertable)){
      trait_table <- paste(db_fp, 
      	"default_traits_img_precalculated.txt.gz", sep='/')
    } else{
      trait_table <- opts$usertable
    }
  }
} else {
  copy_no_file <- paste(db_fp, "16S_13_5_precalculated.txt.gz", sep='/')
  taxonomy <- paste(db_fp, "97_otu_taxonomy.txt.gz", sep='/')
  if(isTRUE(opts$kegg)){
    trait_table <- paste(db_fp, "kegg_modules_precalculated.txt.gz", sep='/')
    if(!isTRUE(opts$predict)){
      if(is.null(opts$phenotype)){
      stop("A list of modules must be specified when using KEGG")
      }
    }
  } else {
    if(is.null(opts$usertable)){
      trait_table <- paste(db_fp, 
      	"default_traits_precalculated.txt.gz", sep='/')
    } else{
      trait_table <- opts$usertable
    }
  }
}

#Check for 'plot all'
#If not 'plot all', check map and column exist
if(!isTRUE(opts$all)){
  if(isTRUE(opts$predict)){
    map <- NULL
    mapcolumn <- NULL
    groups <- NULL

  } else {
    if(is.null(opts$mappingfile)){
    stop("Mapping file not specified. 
         To run BugBase without a mapping file use '-a'")
    } else {
      map <- opts$mappingfile
    }
    if(is.null(opts$mapcolumn)){
      stop("Column header must be specified")
    } else {
      mapcolumn <- opts$mapcolumn
      groups <- opts$groups
    }
  }
} else {
  map <- NULL
  mapcolumn <- NULL
  groups <- NULL
}

# If continuous, remove groups
if(isTRUE(opts$continuous)){
	groups <- NULL #plotting by continuous ignores groups in the column
}

# Define OTU table
if(is.null(opts$otutable)){
	stop("No otu table specified")
} else {
	otu_table <- opts$otutable
}

# Define threshold
threshold_set <- opts$threshold
if(! is.null(threshold_set)){
	if(! 1 >= threshold_set) {
		stop("Threshold must be between 0 and 1")
	}
	if(! threshold_set > 0){
		stop("Threshold must be between 0 and 1")
	}
}

# Define taxa level
taxa_level <- opts$taxalevel
if(! is.null(taxa_level)){
	if(! taxa_level %in% c(1,2,3,4,5,6,7)){
		stop("Taxa level must be 1,2,3,4,5,6 or 7")
	}
}

# Define trait (phenotype) to predict, default is all
test_trait <- opts$phenotype
if(! is.null(test_trait)){
	if(test_trait == "all"){
		test_trait <- NULL
	}
} else {
	test_trait <- test_trait
}

# Define metric for threshold calculations (variance is default)
use_cov <- opts$cov
if(! isTRUE(use_cov)){
	use_cov <- NULL
}

#Define RA or CLR transform
clr_trans <- opts$clr_trans
if(! isTRUE(clr_trans)){
  clr_trans <- NULL
}

# Make output directories
output <- opts$output
if(output != "."){
	if(!file.exists(output)){
		dir.create(output, showWarnings = FALSE, recursive = TRUE)
		dir.create(file.path(output, "normalized_otus"))
		dir.create(file.path(output, "predicted_phenotypes"))
		dir.create(file.path(output, "thresholds"))
		dir.create(file.path(output, "otu_contributions"))
	} else {
		stop("Output directory already exists")
	}
} else {
	stop("Cannot create a hidden directory")
}

print("Loading Inputs...")
#Load inputs
#Required: otu table
#Options: map, map column,groups
loaded.inputs <- load.inputs(otu_table, map, mapcolumn, groups)

if(isTRUE(opts$shotgun)){
  print("Shotgun specified, no copy number normalization will take place...")
} else {
  print("16S copy number normalizing OTU table...")
  #16S copy normalize otu table
  #Required: copyNo_table, loaded otu
  normalized_otus <- copyNo.normalize.otu(copy_no_file, 
  	loaded.inputs$otu_table, output)
}

if(isTRUE(clr_trans)){
  print("Center log transforming OTU table...")
}

print("Predicting phenotypes...")
#Make predictions
#Required:trait table,  normalized otu table
#Options: single trait, threshold, use cov
if(isTRUE(opts$shotgun)){
  prediction_outputs <- single.cell.predictions(trait_table, 
                                                loaded.inputs$otu_table, 
                                                test_trait,
                                                threshold_set,
                                                use_cov, 
                                                clr_trans)
} else {
  prediction_outputs <- single.cell.predictions(trait_table, 
                                                normalized_otus, 
                                                test_trait,
                                                threshold_set,
                                                use_cov,
                                                clr_trans)
}

if(isTRUE(opts$predict)){
  print("BugBase analysis complete")
} else {
  print("Plotting thresholds...")
  #Plot thresholds
  #Two options - one with no mapping file, one with mapping file
  if(is.null(threshold_set)){
    if(isTRUE(opts$all)){
      #Required: predictions
      plot.thresholds.all(prediction_outputs$predictions, clr_trans)
    } else {
      #Required: predictions, map and map column
      plot.thresholds(prediction_outputs$predictions, 
                    loaded.inputs$map, 
                    loaded.inputs$map_column, 
                    clr_trans)
    }
  }

  print("Plotting predictions...")
  #Plot predictions
  #Three options: one without a mapping file, 
  #   one with a mapping file, continous
  #   one with a mapping file, discrete
  if(isTRUE(opts$all)){
    #Required: predictions
    plot.predictions.all(prediction_outputs$final_predictions, clr_trans)
  } else {
    if(isTRUE(opts$continuous)){
      #Required: predictions, map, map column
      plot.predictions.continuous(prediction_outputs$final_predictions, 
                                  loaded.inputs$map, 
                                  loaded.inputs$map_column, 
                                  clr_trans)
    } else {
      #Required: predictions, map, map column
      plot.predictions.discrete(prediction_outputs$final_predictions, 
                                loaded.inputs$map, 
                                loaded.inputs$map_column, 
                                clr_trans)
    }
  }

  print("Plotting OTU contributions...")
  #Plot otu contributions (taxa summaries)
  #Two options, with a mapping file or without
  if(isTRUE(opts$shotgun)){
    if(isTRUE(opts$all)){
      #Required: otu contributions, normalized otu table, taxonomy
      otu.contributions.all.r(prediction_outputs$otus_contributing,
                            prediction_outputs$otu_table_subset,
                            taxonomy, taxa_level)
    } else {
      #Required: otu contributions, normalized otu table, taxonomy
      #   map, map column, taxa_level
      otu.contributions(prediction_outputs$otus_contributing, 
                        prediction_outputs$otu_table_subset, 
                        taxonomy, 
                        loaded.inputs$map, 
                        loaded.inputs$map_column,
                        taxa_level)
    }
  } else {
    if(isTRUE(opts$all)){
      #Required: otu contributions, normalized otu table, taxonomy
      otu.contributions.all.r(prediction_outputs$otus_contributing,
                            prediction_outputs$otu_table_subset,
                            taxonomy, taxa_level)
    } else {
      #Required: otu contributions, normalized otu table, taxonomy
      #   map, map column, taxa_level
      otu.contributions(prediction_outputs$otus_contributing, 
                        prediction_outputs$otu_table_subset, 
                        taxonomy, 
                        loaded.inputs$map, 
                        loaded.inputs$map_column,
                        taxa_level)
    }
  }
print("BugBase analysis complete")
}

File: script/BugBase/bin/run.bugbase.web.py
#!/usr/bin/env python

# Runs BugBase analysis:
# - Uses a single-cell approach to predict microbiome phenotypes
# - plots phenotype relative abundances
# - prints statisical analyses
# - plots thresholds
# - plots otu contributions

# Options
# -t    Specify which taxa level to plot otu contributions by (number 1-7) 
# -p    Specify list of traits (phenotypes) to test (list, comma separated)
# -T    Specify a threshold to use for all traits (number 0-1)
# -g    Specify subset of groups in map column to plot (list, comma separated)
# -u  Use a user-define trait table. Absolute file path must be specified
# -k  Use the kegg pathways instead of default traits
# -z    Data is of type continuous 
# -C    Use covariance instead of variance to determine thresholds
# -a    Plot all samples (no stats will be run)
import site
import sys
import os
import random
import operator
import csv
from subprocess import Popen, PIPE, STDOUT
from optparse import OptionParser

# These are the environment paths added by the module:

# Python
os.environ['PATH'] = os.environ['PATH'] + '/soft/python-2.7/bin'

# PyCogent
os.environ['PYTHONPATH'] = ':/web/research/bugbase.cs.umn.edu/site-packages/PyCogent-1.5.3'

# R
#os.environ['PATH'] = os.environ['PATH'] + ':/soft/r/3.1.2/linux_x86_64/bin'
#os.environ['PATH'] = os.environ['PATH'] + ':/soft/r/3.1.2/linux_x86_64/rstudio-0.98.1087/bin'
os.environ['PATH'] = os.environ['PATH'] + ':/soft/r/3.2.2/linux_x86_64/bin'
os.environ['PATH'] = os.environ['PATH'] + ':/soft/r/3.2.2/linux_x86_64/rstudio-0.99.486/bin'

# BugBase
os.environ['PATH'] = os.environ['PATH'] + ':/web/research/bugbase.cs.umn.edu/BugBase/bin'
os.environ['BUGBASE_PATH'] = '/web/research/bugbase.cs.umn.edu/BugBase/'

def make_option_parser():
	parser = OptionParser(usage="usage: %prog [options] filename",
		version="%prog 1.0")
	parser.add_option("-v","--verbose",
		action="store_true",
		default=True,
		help="Verbose output (default %default)",)
	parser.add_option("-p","--print_only",
		action="store_true",
		default=False,
		help="Only print commands (default %default)",)
	parser.add_option("-i", "--input_OTU",
		default=None,
		type='string',
		help="OTU table (required)")
	parser.add_option("-m", "--mapping_file",
		default=None,
		type='string',
		help="mapping file (required)")
	parser.add_option("-o", "--output",
		default=".",
		type='string',
		help="Output directory (default %default)")
	parser.add_option("-c","--map_column",
		default=None,
		type='string',
		help="name of column that lists treatment groups")
	parser.add_option("-P","--phenotypes",
		default=None,
		type='string',
		help="specific phenotypes to predict, comma separated list, no spaces")
	parser.add_option("-x","--predict",
		default=False,
		action="store_true",
		help="only output the prediction table, do not make plots")
	parser.add_option("-T","--threshold",
		default=None,
		type='float',
		help="threshold (0 to 1) you would like to set for the trait listed")
	parser.add_option("-g","--groups",
		default=None,
		type='string',
		help="treatment groups you would like to plot, separated by commas with no spaces")
	parser.add_option("-t","--taxalevel",
		default=None,
		type='float',
		help="taxa level to plot otu contributions by, default is 2 (phylum)")
	parser.add_option("-a","--plot_all",
		action="store_true",
		default=False,
		help="Plot all samples without treatment-group seperation and no statistics")
	parser.add_option("-C","--cov",
		action="store_true",
		default=False,
		help="use coefficient of variance instead of variance")
	parser.add_option("-l","--clr",
		action="store_true",
		default=False,
		help="use centered log-ratio transformation instead of relative abundance")
	parser.add_option("-k","--kegg",
		action="store_true",
		default=False,
		help="use kegg pathway table")
	parser.add_option("-z","--continuous",
		action="store_true",
		default=False,
		help="Plot data according to a continuous variable (picked against IMG)")
	parser.add_option("-w","--wgs",
		action="store_true",
		default=False,
		help="Data is whole genome shotgun data")
	return parser

def run_commands(commands, print_only=False, verbose=True, error_on_fail=True):
	return_vals = []

	# Run all commands
	for cmd in commands:
		print cmd
		if not print_only:
			proc = Popen(cmd,shell=True,universal_newlines=True,stdout=PIPE,stderr=PIPE)
			stdout, stderr = proc.communicate()

		# If requested, prints all output from the program
			if verbose:
				print stdout
				print stderr
			if error_on_fail == True and proc.returncode != 0:
				print stdout
				print stderr
				raise ValueError('Command failed: ' + cmd)

			return_vals.append(proc.returncode)
	return(return_vals)

		
if __name__ == '__main__':

	parser = make_option_parser()
	(options, args) = parser.parse_args()

	# Name user inputs
	output = " -o /web/research/bugbase.cs.umn.edu/results/" + options.output

	otu_table = " -i /web/research/bugbase.cs.umn.edu/uploads/" + options.input_OTU 
	
	if options.mapping_file:
		mapping = " -m /web/research/bugbase.cs.umn.edu/uploads/" + options.mapping_file
		map_file = "/web/research/bugbase.cs.umn.edu/uploads/" + options.mapping_file
	else:
		mapping = ""
		
	if options.map_column:
		column = " -c " + options.map_column
		col_name = options.map_column
	else:
		column = ""

	if options.phenotypes:
		phenos = " -p " + options.phenotypes
	else:
		phenos = ""

	if options.predict:
		predict_only = " -x"
	else:
		predict_only = ""

	if options.threshold:
		thres = " -T %s" %(options.threshold)
	else:
		thres = ""

	if options.groups:
		groups = " -g " + options.groups
		groups_list = options.groups.split(",")
	else:
		groups = ""

	if options.taxalevel:
		taxa = " -t %s" %(options.taxalevel)
	else:
		taxa = ""

	if options.plot_all is True:
		plot_all = " -a"
	else:
		plot_all = ""

	if options.cov is True:
		cov = " -C"
	else:
		cov = ""

	if options.clr is True:
		clr = " -l"
	else:
		clr = ""

	if options.kegg is True:
		kegg = " -k"
	else:
		kegg = ""

	if options.continuous is True:
		continuous = " -z"
	else:
		continuous = ""

	if options.wgs is True:
		wgs = " -w"
	else:
		wgs = ""

	if options.plot_all is False:
		if options.predict is False:
			if options.mapping_file is None:
				print "[ERROR_MESSAGE]Mapping file must be specified"
				sys.exit()
			if options.map_column is None:
				print "[ERROR_MESSAGE]Column header must be specified"
				sys.exit()

			# Make sure map column is valid
			with open(map_file, 'rU') as input_map:
				reader = csv.reader(input_map, delimiter='\t')
				headers = reader.next()
			if col_name in headers:
				print col_name + " was specified as map column header\n"
			else:
				print "[ERROR_MESSAGE]Column header specified does not exist in mapping file\n"
				print "[ERROR_MESSAGE]These are the available column headers: "+ ', '.join(headers)
				sys.exit()
		
			# If groups are specified, check they are valid
			if options.groups is not None:
				groups_avail = []
				with open(map_file, 'rU') as input_map:
					reader = csv.reader(input_map, delimiter='\t')
					headers = reader.next()
					column_index = headers.index(col_name)
					for row in reader:
						name = str(row[column_index])
						groups_avail.append(name)   
				for group_defined in groups_list:
					if group_defined in groups_avail:
						if len(groups_list) <= 1:
							print "[ERROR_MESSAGE]A minimum of two groups must be tested"
							sys.exit()
					else:
						groups_avail = list(set(groups_avail))
						print "[ERROR_MESSAGE]Groups specified do not exist in mapping file"
						print "[ERROR_MESSAGE]These are the groups available under " + col_name + " header: " + ', '.join(groups_avail)
						sys.exit()
	
	# If threshold is user-specified, state what will be used
	#if options.threshold is not None:
	#	print "A user-specified threshold of %s will be used for all traits" %(options.threshold)

	# If kegg is specified, check that modules are valid
	if options.kegg is True:
		if options.predict is False:
			if options.phenotypes is None:
				print "[ERROR_MESSAGE]When using KEGG, modules to plot must be specified with a comma separated list"

	commands = []
						
	# Run the run.bugase.web.r script
	
	#cmd = "/soft/r/3.2.2/linux_x86_64_old/bin/Rscript /web/research/bugbase.cs.umn.edu/BugBase/bin/run.bugbase.web.r -h" 
	cmd = "Rscript /web/research/bugbase.cs.umn.edu/BugBase/bin/run.bugbase.web.r" + otu_table + mapping + column + phenos + predict_only + taxa + groups + thres + plot_all + cov + clr + kegg + wgs + continuous + output 
	#cmd = "/soft/r/3.2.2/linux_x86_64_old/bin/Rscript /web/research/bugbase.cs.umn.edu/BugBase/bin/test.r" + otu_table + mapping + column + phenos + predict_only + taxa + groups + thres + plot_all + cov + kegg + wgs+ continuous + output
	commands.append(cmd)
	
	# Run commands
	return_vals = run_commands(commands, print_only=options.print_only, verbose=options.verbose)
		
	print "[SUCCESSFUL] Data is analyzed"
